gordon_janitor_gcp/gce_clients.py

```diff
@@ -0,0 +1,297 @@
+            tagged with any of these.
+        blacklisted_metadata (list): Do not collect an instance if its metadata
+            key:val matches a {key:val} dict in this list.
+    """
+    BASE_URL = 'https://www.googleapis.com/compute/v1/projects'
+
+    def __init__(self,
+                 auth_client=None,
+                 session=None,
+                 blacklisted_tags=None,
+                 blacklisted_metadata=None):
+        super().__init__(auth_client=auth_client, session=session)
+        self.blacklisted_tags = blacklisted_tags or []
+        self.blacklisted_metadata = blacklisted_metadata or []
+
+    async def list_instances(self,
+                             project,
+                             page_size=500,
+                             instance_filter=None,
+                             retries=3,
+                             retry_wait=5):
+        """Fetch all instances in a GCE project.
+
+        Endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Args:
+            project (str): unique, user-provided project ID.
+            page_size (int): hint for the client to only retrieve up to this
+                number of results per API call.
+            instance_filter (str): endpoint-specific filter string used to
+                retrieve a subset of instances. This is passed directly to the
+                endpoint's "filter" URL query parameter.
+        Returns:
+            list: dicts containing instance data.
+        """
+        url = f'{self.BASE_URL}/{project}/aggregated/instances'
+        params = {'maxResults': page_size}
+        if instance_filter:
+            params['filter'] = instance_filter
+
+        return await self.list_all_items(url, params, retries, retry_wait)
+
+    def parse_response_into_items(self, response, items):
+        """Parse API response for all instances in a project into a list.
+
+        Args:
+            response (dict): endpoint response data.
+            items (list): list for collecting parsed results.
+        """
+        zones = response.get('items', {})
+        for zone in zones.values():
+            for instance in zone.get('instances', []):
+                if not any([
+                        self._blacklisted_by_tag(instance),
+                        self._blacklisted_by_metadata(instance)
+                ]):
+                    try:
+                        items.append(self._extract_instance_data(instance))
+                    except KeyError as e:
+                        logging.error(
+                            'Could not extract instance information for '
+                            f'{instance} because of missing key {e}, skipping.')
+
+    def _extract_instance_data(self, instance):
+        return {
+            'hostname': instance['name'],
+            'internal_ip': instance['networkInterfaces'][0]['networkIP'],
```

Comment:
I used square brackets in this instances because I wanted to catch, log, and ignore errors where incomplete instance data was returned, but if someone knows a better approach then I'm game.